### PR TITLE
feat(clients/go): set variables on fail job

### LINKED ISSUE #10788: Set variables on Fail Job via Go client
**Is your feature request related to a problem? Please describe.**
Recently, we introduced the ability to set variables when failing a job to Zeebe. The FailJob RPC now supports passing these variables along. Additionally, the `zeebe-java-client` now provides an interface to pass variables along when failing a job, but the Go client was left out of scope.
- #9404 

**Describe the solution you'd like**
Add a way to pass variables along when failing a job via the Go client. Inspiration can be taken from other RPCs that support passing variables along like `CompleteJob`.

**Describe alternatives you've considered**
It's possible to add this support to zbctl once it's implemented in the Go client, but I don't see a need for it now. We can implement this in a separate issue if/when users ask for it.

**Additional context**
This functionality was implemented for the Java client in:
- #10705

### PR DESCRIPTION
**Pull request by lzgabel**
## Description

Set variables on fail job via go client.

## Related issues

closes #10788 

## Definition of Done

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by lzgabel** [comment]
🆘 Hi all. I need help on how to fix `.gocompat.json` file when there is a breaking change in the go client API. 

I tried running `gocompat save ./...` as mentioned in this [wiki](https://github.com/camunda/zeebe/wiki/Breaking-Changes#go), but it doesn't work.

File: clients/go/cmd/zbctl/internal/commands/failJob.go [hunk 1]
```diff
@@ -40,6 +40,7 @@ var (
 	failJobRetriesFlag      int32
 	failJobRetryBackoffFlag time.Duration
 	failJobErrorMessage     string
+	failJobVariablesFlag    string
```
Comment (korthout) [inline clients/go/cmd/zbctl/internal/commands/failJob.go:43]: AFAIK this isn't a flag

🔧 I think we can rename this to `failJobVariables`

**Review by korthout — COMMENTED** [review]
Thanks for the changes @lzgabel 👍 and sorry for the long wait 🙇 

💯 The changes themselves look good to me. Nice work!

>I tried running gocompat save ./... as mentioned in this [wiki](https://github.com/camunda/zeebe/wiki/Breaking-Changes#go), but it doesn't work.

Can you tell me what happens? In what way does it not work?

**Comment by lzgabel** [comment]
> Can you tell me what happens? In what way does it not work?

Hi @korthout. It seems that I solved it by reinstalling `gocompat`, please check this out. 🙇

File: clients/go/cmd/zbctl/internal/commands/failJob.go (hunk 1 above)
Comment (lzgabel) [reply clients/go/cmd/zbctl/internal/commands/failJob.go]: > AFAIK this isn't a flag

❓ Hi @korthout. I have a question. Is this different from `completeJob` ? 

https://github.com/camunda/zeebe/blob/7340a588b972bae4d1e33ae04f6673619324440d/clients/go/cmd/zbctl/internal/commands/completeJob.go#L36-L39

Comment (korthout) [reply clients/go/cmd/zbctl/internal/commands/failJob.go]: Ah, that's interesting. My understanding of what a flag is in the context of Cobra appears to be incorrect.

I was expecting flags to be either boolean or not, but in Cobra each argument/option of a command is called a flag. See, https://cobra.dev/#flags.

🔧 Feel free to re-add the flag suffix. But it's fine either way.

Comment (lzgabel) [reply clients/go/cmd/zbctl/internal/commands/failJob.go]: It's okay, let's merge this now.

**Comment by korthout** [comment]
@lzgabel There appears to be [a failing test](https://github.com/camunda/zeebe/actions/runs/3363439635/jobs/5586631356):

>```
> === RUN   TestFailJobCommand_VariablesFromObjectIgnoreOmitempty
>     fail_job.go:114: Unexpected call to *mock_pb.MockGatewayClient.FailJob([context.Background.WithDeadline(2022-11-01 09:32:17.777604347 +0000 UTC m=+5.010333110 [4.999874898s]) jobKey:123 retries:12 errorMessage:"something went wrong" variables:"{}"]) at /home/runner/work/zeebe/zeebe/clients/go/pkg/commands/fail_job.go:114 because: 
>         expected call at /home/runner/work/zeebe/zeebe/clients/go/pkg/commands/fail_job_test.go:298 doesn't match the argument at index 1.
>         Got: jobKey:123 retries:12 errorMessage:"something went wrong" variables:"{}" (*pb.FailJobRequest)
>         Want: is jobKey:123 retries:12 errorMessage:"something went wrong" variables:"{\"foo\":\"\"}"
>     controller.go:269: missing call(s) to *mock_pb.MockGatewayClient.FailJob(is anything, is jobKey:123 retries:12 errorMessage:"something went wrong" variables:"{\"foo\":\"\"}") /home/runner/work/zeebe/zeebe/clients/go/pkg/commands/fail_job_test.go:298
>     controller.go:269: aborting test due to missing call(s)
> ```

Please also make sure to run [gofmt with -s](https://github.com/camunda/zeebe/actions/runs/3363439635/jobs/5586631877)

**Comment by lzgabel** [comment]
Hi @korthout. Please try it again. 🙇

**Comment by lzgabel** [comment]
Hi @korthout. I've fixed the failing test. Please check this out. 🙇

**Review by korthout — APPROVED** [review]
Thanks @lzgabel this looks great 🚀 

bors merge

**Pull request merged**